FDC3 For Web Browsers Implementation

robmoffat opened the pull request:
Implementation PR for FDC3 for the Web (#896)

- resolves #1209 
- resolves #1297 
- resolves #1429 
- resolves #1430
- resolves #1431
- resolves #1432 
- resolves #1433
- resolves #1468
- resolves #810
- resolves #832
- resolves #1487
- resolves #1488

This PR primarily implements the `getAgent()` function (in the [_**fdc3-get-agent**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-get-agent) package) as defined by the [FDC3 Web Connection Protocol (WCP)](https://fdc3.finos.org/docs/next/api/specs/webConnectionProtocol) and a `DesktopAgentProxy` (in the [_**fdc3-agent-proxy**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-agent-proxy) package) that it uses to implement communication with browser-based Desktop Agents via the [FDC3 Desktop Agent Communication Protocol (DACP)](https://fdc3.finos.org/docs/next/api/specs/desktopAgentCommunicationProtocol).

Additionally, the FDC3 repository is refactored to be a monorepo supported by `npm workspaces`. The existing repository was split into [_**fdc3-context**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-context) (context type definitions), [_**fdc3-schems**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-schema) (bridging and DACP protocol messages) and [_**fdc3-standard**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-standard) (existing API TypeScript definitions, Intents and App Directory) packages. A [_**testing**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/testing) package was added with tooling used in multiple packages for testing.

These packages are rolled up for release on NPM by an [_**fdc3**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3) package (publishing to the existing [@finos/fdc3](https://www.npmjs.com/package/@finos/fdc3) NPM package). Previous releases provided both ES6 and CommonJs modules within a single package - this release will only provide ES6, with a separate [_**fdc3-commonjs**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/packages/fdc3-commonjs) module provided for backwards compatibility (will release at [@finos/fdc3-commonjs](https://www.npmjs.com/package/@finos/fdc3-commonjs)).

Finally, a reference implementation, demo for FDC3 in a Web Browser and reference channel selector and intent resolver UIs (that can be injected into an app by getAgent) are provided in the [_**toolbox/fdc3-for-web**_](https://github.com/finos/FDC3/tree/fdc3-for-web-impl/toolbox/fdc3-for-web) package.

julianna-ciq commented:
@robmoffat  Please see https://github.com/finos/FDC3/pull/1405 for build fixes

julianna-ciq commented on packages/fdc3-agent-proxy/test/step-definitions/channels.steps.ts:
```diff
@@ -0,0 +1,179 @@
+
+Given("channel {string} has context {string}", function (this: CustomWorld, channel: string, context: string) {
+  const ctxObject = handleResolve(context, this)
+  const state = this.props[CHANNEL_STATE] ?? {}
+  this.props[CHANNEL_STATE] = state
```
```suggestion
  this.props[CHANNEL_STATE] ??= {}
  const state = this.props[CHANNEL_STATE]
```

julianna-ciq commented on packages/fdc3-agent-proxy/test/step-definitions/channels.steps.ts:
```diff
@@ -0,0 +1,179 @@
+  const state = this.props[CHANNEL_STATE] ?? {}
+  this.props[CHANNEL_STATE] = state
+
+  const cs = state[channel] ?? []
+  cs.push(ctxObject)
+  state[channel] = cs
```
```suggestion
  state[channel] ??= []
  state[channel].push(ctxObject)
```

julianna-ciq reviewed (changes requested):
Reviewing the addon and fdc3-agent-proxy package. More reviews to come...

julianna-ciq commented on packages/fdc3-get-agent/src/strategies/ElectronEventLoader.ts:
```diff
@@ -0,0 +1,36 @@
+    get(params: GetAgentParams): Promise<DesktopAgent | void> {
+        return new Promise<DesktopAgent | void>((resolve, reject) => {
+            const endPollTime = Date.now() + (params.timeoutMs!! + 500)
+            // console.log("Starting poll: " + endPollTime + " " + params.timeout + " " + Date.now())
```
Did you mean to leave this line here?

julianna-ciq replied:
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/strategies/PostMessageLoader.ts:
```diff
@@ -0,0 +1,160 @@
+    return ifrm.contentWindow!!
+}
+
+export function helloExchange(options: GetAgentParams, connectionAttemptUuid: string): Promise<ConnectionDetails> {
```
What happens to this method if data.type doesn't equal `WCP3Handshake`? It looks like it won't resolve otherwise... do you just let it hang and timeout? What's the expected interaction?

julianna-ciq commented on packages/fdc3-get-agent/src/strategies/TimeoutLoader.ts:
```diff
@@ -0,0 +1,35 @@
+
+    get(params: GetAgentParams): Promise<DesktopAgent | void> {
+        return new Promise<DesktopAgent | void>((resolve, reject) => {
+            const endPollTime = Date.now() + params.timeoutMs!!
```
```suggestion
            const endPollTime = Date.now() + params.timeoutMs
```
Handle null values

julianna-ciq replied:
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/strategies/getAgent.ts:
```diff
@@ -0,0 +1,114 @@
+        })
+        .catch(async (error) => {
+            if (options.failover) {
+                const o = await handleWindowProxy(options, () => { return options.failover!!(options) })
+                return o
```
```suggestion
                return await handleWindowProxy(options, () => { return options.failover?.(options) })
```

julianna-ciq replied:
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/ui/AbstractUIComponent.ts:
```diff
@@ -0,0 +1,127 @@
+        port.addEventListener("message", (e) => {
+            const data = e.data
+            if (data.type == 'Fdc3UserInterfaceRestyle') {
+                // console.log(`Restyling ${JSON.stringify(data.payload)}`)
```
Did you mean to leave this in place?

julianna-ciq commented on packages/fdc3-get-agent/src/ui/AbstractUIComponent.ts:
```diff
@@ -0,0 +1,127 @@
+    private awaitHello(): Promise<MessagePort> {
+        return new Promise((resolve, _reject) => {
+            const ml = (e: MessageEvent) => {
+                // console.log("Received UI Message: " + JSON.stringify(e.data))
```
Did you mean to leave this in place?

julianna-ciq replied:
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/ui/DefaultDesktopAgentChannelSelector.ts:
```diff
@@ -0,0 +1,58 @@
+    private callback: ((channelId: string | null) => void) | null = null
+
+    constructor(url: string | null) {
+        super(url ?? "https://fdc3.finos.org/webui/channel_selector.html", "FDC3 Channel Selector")
```
It'd be nice if this was stored as a constant somewhere

julianna-ciq commented on packages/fdc3-get-agent/src/ui/DefaultDesktopAgentChannelSelector.ts:
```diff
@@ -0,0 +1,58 @@
+
+    updateChannel(channelId: string | null, availableChannels: Channel[]): void {
+        // also send to the iframe
+        this.port!!.postMessage({
```
```suggestion
        this.port?.postMessage({
```

julianna-ciq replied:
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/ui/NullIntentResolver.ts:
```diff
@@ -0,0 +1,16 @@
+import { IntentResolver, IntentResolutionChoice } from "@kite9/fdc3-standard";
+import { Context } from "@kite9/fdc3-context";
+
+
```
Can you add a comment about what this file is doing, why it exists? Is it used for testing, for development, is it publicly facing?

julianna-ciq commented on packages/fdc3-get-agent/test/features/default-channel-selector.feature:
```diff
@@ -0,0 +1,20 @@
+  Background: Desktop Agent API
+    Given a browser document in "document" and window in "window"
+    And A Channel Selector in "channel-selector" with callback piping to "cb"
+    Given User Channels one, two and three in "channel-list"
```
```suggestion
    And User Channels one, two and three in "channel-list"
```

julianna-ciq commented on packages/fdc3-get-agent/test/features/default-channel-selector.feature:
```diff
@@ -0,0 +1,20 @@
+    Given User Channels one, two and three in "channel-list"
+
+  Scenario: Channel Selector Requests Channel Change
+    Given The channel selector sends a channel change message for channel "one"
```
I think this should be a "When"

julianna-ciq commented on packages/fdc3-get-agent/test/features/default-channel-selector.feature:
```diff
@@ -0,0 +1,20 @@
+    And I call "{document}" with "shutdown"
+
+  Scenario: Updating channel information in the channel selector
+    Given I call "{channel-selector}" with "updateChannel" with parameters "one" and "{channel-list}"
```
I think "When" is the more appropriate verb here.

julianna-ciq commented on packages/fdc3-get-agent/test/features/default-intent-resolver.feature:
```diff
@@ -0,0 +1,29 @@
+    Given I call "{intent-resolver}" with "chooseIntent" with parameters "{appIntents}" and "{instrumentContext}" for a promise
+    And I refer to "{result}" as "theIntentPromise"
+    And we wait for a period of "200" ms
+    Given The intent resolver sends an intent selection message
```
I think this should be a When

julianna-ciq commented on packages/fdc3-get-agent/test/features/default-intent-resolver.feature:
```diff
@@ -0,0 +1,29 @@
+    Given I call "{intent-resolver}" with "chooseIntent" with parameters "{appIntents}" and "{instrumentContext}" for a promise
+    And I refer to "{result}" as "theIntentPromise"
+    And we wait for a period of "200" ms
+    Given The intent resolver cancels the intent selection message
```
I think this should be a When

julianna-ciq commented on packages/fdc3-get-agent/test/support/TestServerContext.ts:
```diff
@@ -0,0 +1,156 @@
+
+            this.instances.push(connectionDetails)
+            internalPort.onmessage = (msg) => {
+                console.log(`Received message on internalPort ${appId}: ${JSON.stringify(msg.data)}`)
```
Did you intend to leave this console.log line? Either add a comment to be explicit, or use console.debug

julianna-ciq replied:
Done in #1424

julianna-ciq commented on packages/fdc3-get-agent/test/support/MockWindow.ts:
```diff
@@ -0,0 +1,70 @@
+
+    addEventListener(type: string, callback: (e: Event) => void): void {
+        this.eventHandlers.push({ type, callback });
+        console.log("Added event handler " + this.tag);
```
Did you mean to leave this line here?

julianna-ciq commented on packages/fdc3-get-agent/test/step-definitions/desktop-agent.steps.ts:
```diff
@@ -0,0 +1,152 @@
+})
+
+After(function (this: CustomWorld) {
+    console.log("Cleaning up")
```
Should this line stay?

julianna-ciq replied:
Done in #1424

julianna-ciq commented on packages/fdc3-get-agent/test/step-definitions/desktop-agent.steps.ts:
```diff
@@ -0,0 +1,152 @@
+After(function (this: CustomWorld) {
+    console.log("Cleaning up")
+    clearAgentPromise()
+    // setTimeout(() => {
+    //     //console.log((process as any)._getActiveHandles())
+    //     wtf.dump()
+    // }, 10000)
```
Should these lines stay?

julianna-ciq replied:
Done in #1424

julianna-ciq commented on packages/fdc3-get-agent/test/step-definitions/intent-resolver.steps.ts:
```diff
@@ -0,0 +1,126 @@
+    })
+})
+
+// Given('{string} receives a {string} message for the intent resolver and pipes comms to {string}', async function (this: CustomWorld, frame: string, type: string, output: string) {
```
If rthese lines aren't needed, they should be deleted

julianna-ciq commented on packages/fdc3-get-agent/test/support/MockFDC3Server.ts:
```diff
@@ -0,0 +1,100 @@
+                const source = event.source as Window
+                const origin = event.origin;
+
+                console.log("Received: " + JSON.stringify(event.data));
```
Did you mean to leave this line here?

julianna-ciq reviewed (commented):
Reviews for the fdc3-get-agent package

kriswest reviewed (changes requested):
Just starting a review - have only really got through addon

kriswest commented:
@robmoffat you'll need to merge fdc3-for-web again, bunch of minior corrections, mostly in schema comments. SHouldn't affect actual code so hopefully will just merge.

robmoffat commented:
> @robmoffat you'll need to merge fdc3-for-web again, bunch of minior corrections, mostly in schema comments. SHouldn't affect actual code so hopefully will just merge.

ok.  Will do.  

Also, there's zero point in reviewing this branch as-is:  I know your time is limited on FDC3 right now so  I'll let you know when you should take a look at it. But it's not yet.

robmoffat commented on packages/addon/index.html:
```diff
@@ -0,0 +1,61 @@
+<!-- <!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Channel Selector</title>
+  </head>
+  <body data-expanded="true">
+    <div id="list"></div>
+    <div id="tokens"></div>
+    <script type="module" src="/src/main.ts"></script>
+  </body>
+</html> -->
+
```
you're going to do this though, right?

julianna-ciq replied:
Done in #1424

kriswest commented:
when I say merge, I probably mean rebase...

robmoffat commented:
Can we stop "reviewing" this PR?  There's plenty of work left to sort out in the implementation (@julianna-ciq have a look at https://github.com/finos/FDC3/issues/1297 as well as all the disabled tests) before we start reviewing it.  At the moment this feels like complaining that there's a big pile of timber in your remodelled living room... while the carpenters are still in!

On the @kite9 / @finos thing - I don't think we can really change this until we're ready to do a release, and that really needs both #1297 and #1191 to be completed first.

kriswest commented:
@robmoffat Apologies, we've been trying to work out how and where to contribute to getting this into a complete and production-ready state for a while. 

There's no readme, description on the PR, todo list or timeline at the moment and the PR is vast (as it combines new implementation with a wholesale refactor of the repository - something that might have been more easily accomplished as separate PRs). I'm regularly asked when this will be complete and I have no idea at present. The only thing we could do is start looking at it and to try and work out what's needed where while we try and arrange time with you to get some guidance. In the meantime, we can remove those demo apps and look at the docusaurus build (2 / 4 remaining items on the list you pointed to on #1297 - presumably that list only relates to #1297 and not #896 directly, we weren't aware of it until you referred to it yesterday).

I don't see how #1297 (mono repo refactor) can be completed before this is as it's part of this PR currently. It is understood that we can't change the npm module names and their use as deps until release, but we typically mark out things in review that need to be ticked off (conversation resolved) before release. 

#1197 is awaiting final reviews (as all comments raised are now resolved), hopefully at least two of @kemerava @bingenito @hughtroeger and/or @novavi will be able to sign that off soon so it can merge.

robmoffat commented:
> I don't see how https://github.com/finos/FDC3/issues/1297 (mono repo refactor) can be completed before this is as it's part of this PR currently. It is understood that we can't change the npm module names and their use as deps until release, but we typically mark out things in review that need to be ticked off (conversation resolved) before release.

So as a bit of history, this PR got created when Alex and I started work on #1297 (as referenced in the opening comment).  Originally, the implementation of FDC3 for the web was in a finos-labs project and we brought it in here as part of that.  But yes I agree - since this is the implementation of #1297 we can't close that until this is done.

robmoffat commented:
> https://github.com/finos/FDC3/issues/1197 is awaiting final reviews (as all comments raised are now resolved), hopefully at least two of @kemerava @bingenito @hughtroeger and/or @novavi will be able to sign that off soon so it can merge.

I pinged Derek about this today, fwiw.

kriswest commented:
> But yes I agree - since this is the implementation of #1297 we can't close that until this is done.

At least it's helping me persuade myself to buy a bigger laptop (rendering the diff takes some doing) ;-)

It's also a huge contribution of work to the standard which will have a massive positive impact on it if you/we pull it off. Looking forward to helping where get that done.

julianna-ciq commented on package.json:
```diff
@@ -42,33 +48,11 @@
-  "optionalDependencies": {
-    "@rollup/rollup-linux-x64-gnu": "4.14.1"
+  "devDependencies": {
+    "concurrently": "^8.2.2",
```
Done in #1424

julianna-ciq commented on packages/addon/package.json:
```diff
@@ -0,0 +1,19 @@
+  "scripts": {
+    "dev": "vite",
+    "build": "tsc && vite build",
+    "clean": "rimraf dist",
```
Done in #1424

julianna-ciq commented on packages/addon/src/intent_resolver.ts:
```diff
@@ -0,0 +1,215 @@
+  document.body.setAttribute("data-visible", "true");
+  document.querySelector("dialog")?.showModal();
+
+  console.log("setup data:", data);
```
Done in #1424

julianna-ciq commented on packages/addon/src/intent_resolver.ts:
```diff
@@ -0,0 +1,215 @@
+    });
+  });
+
+  console.log(document.body);
```
Done in #1424

julianna-ciq commented on packages/addon/src/intent_resolver_orig.js:
Done in #1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/channels/DefaultChannel.ts:
```diff
@@ -0,0 +1,78 @@
+            },
+            type: "broadcastRequest"
+        } as BroadcastRequest, 'broadcastResponse')
+        console.log("broadcast done", done)
```
Done in #1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/channels/DefaultChannelSupport.ts:
```diff
@@ -0,0 +1,154 @@
+    protected userChannels: Channel[] | null = null
+    private followingListeners: FollowingContextListener[] = []
+
+    constructor(messaging: Messaging, channelSelector: ChannelSelector) {
+        this.messaging = messaging;
+        this.channelSelector = channelSelector
```
Done in #1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/channels/DefaultPrivateChannel.ts:
```diff
@@ -0,0 +1,89 @@
+            }
+        }
+
+        if (type) {
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/intents/DefaultIntentSupport.ts:
```diff
@@ -0,0 +1,177 @@
+type RaiseIntentResponse = BrowserTypes.RaiseIntentResponse
+type FindIntentRequest = BrowserTypes.FindIntentRequest
+
+function convertIntentResult(m: RaiseIntentResultResponse, messaging: Messaging): Promise<IntentResult> {
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/intents/DefaultIntentSupport.ts:
```diff
@@ -0,0 +1,177 @@
+            throw new Error(ResolveError.NoAppsFound)
+        } else {
+            return {
+                intent: appIntent.intent as any /* ISSUE: 1295 */,
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/intents/DefaultIntentSupport.ts:
```diff
@@ -0,0 +1,177 @@
+            payload: {
+                context
+            },
+            meta: this.messaging.createMeta() as AddContextListenerRequestMeta /* ISSUE: #1275 */
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/listeners/AbstractListener.ts:
```diff
@@ -0,0 +1,67 @@
+            }
+        }
+
+        const response = await this.messaging.exchange<any>(notificationMessage, responseType!!)
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/listeners/AbstractListener.ts:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    async register() {
+        const id = await this.listenerNotification(this.subscribeType)!!
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/listeners/PrivateChannelEventListener.ts:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    filter(m: BroadcastEvent) {
+        return (m.type == EVENT_NAMES[this.listenerType]) && (this.privateChannelId == (m.payload as any)?.privateChannelId); /* ISSUE: 1293 */
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts:
```diff
@@ -0,0 +1,152 @@
+        this.timeout = timeout
+    }
+
+    getSource(): AppIdentifier {
+        return this.appId!!
+    }
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts:
```diff
@@ -0,0 +1,152 @@
+                setTimeout(() => {
+                    this.unregister(id)
+                    if (!done) {
+                        console.error(`Rejecting after ${this.timeout}ms with ${timeoutErrorMessage}`)
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts:
```diff
@@ -0,0 +1,152 @@
+
+    async exchange<X>(message: any, expectedTypeName: string, timeoutErrorMessage?: string): Promise<X> {
+        const errorMessage = timeoutErrorMessage ?? `Timeout waiting for ${expectedTypeName} with requestUuid ${message.meta.requestUuid}`
+        const prom = this.waitFor(m =>
+            (m.type == expectedTypeName)
+            && (m.meta.requestUuid == message.meta.requestUuid), errorMessage)
+        this.post(message)
+        const out: any = await prom
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private async exchangeValidationWithId<X>(message: any, connectionAttemptUuid: string): Promise<X> {
+        const prom = this.waitFor(m =>
+            (m.meta.connectionAttemptUuid == connectionAttemptUuid))
+        this.post(message)
+        const out: any = await prom
+        if (out?.payload?.message) {
+            throw new Error(out.payload.message)
+        } else {
+            return out
+        }
```
Done in https://github.com/finos/FDC3/pull/1424

julianna-ciq commented on packages/fdc3-get-agent/src/messaging/MessagePortMessaging.ts:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    waitFor<X>(filter: (m: any) => boolean, timeoutErrorMessage?: string): Promise<X> {
```
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/messaging/MessagePortMessaging.ts:
```diff
@@ -0,0 +1,74 @@
+    waitFor<X>(filter: (m: any) => boolean, timeoutErrorMessage?: string): Promise<X> {
+        // console.log("Waiting for", filter, timeoutErrorMessage)
+        return super.waitFor(filter, timeoutErrorMessage).then((v: any) => {
+            // console.log("Wait over ", v, timeoutErrorMessage)
```
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/messaging/MessagePortMessaging.ts:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    register(l: RegisterableListener): void {
+        this.listeners.set(l.id!!, l)
```
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/messaging/MessagePortMessaging.ts:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    waitFor<X>(filter: (m: any) => boolean, timeoutErrorMessage?: string): Promise<X> {
+        // console.log("Waiting for", filter, timeoutErrorMessage)
```
Fixed in #1424

julianna-ciq commented on packages/fdc3-get-agent/src/messaging/MessagePortMessaging.ts:
```diff
@@ -0,0 +1,74 @@
+        return uuidv4();
+    }
+
+    post(message: object): Promise<void> {
+        this.cd.messagePort.postMessage(message);
+        return Promise.resolve();
+    }
```
Fixed in #1424

julianna-ciq commented on packages/fdc3-context/package.json:
```diff
@@ -0,0 +1,45 @@
+  ],
+  "scripts": {
+    "clean": "rimraf dist && rimraf generated",
+    "mkdirs": "mkdirp generated/context",
```
Fixed in #1424

robmoffat committed:
Converted to constants for all Fdc3UserInterface type usages

robmoffat committed:
Renumbered ports for consistency, added in UI choice again

robmoffat committed:
Fixed demo UI using new FDC3 UI Constants

robmoffat committed:
Got alternative UI components working

robmoffat committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-merging-prs

robmoffat committed:
Merge pull request #1436 from finos/fdc3-for-web-impl-merging-prs

Fdc3 for web impl merging prs

robmoffat committed:
Allowing channel selector drag

robmoffat committed:
Made smoother

kriswest committed:
Linting setup

kriswest committed:
WIP GetAgent refactor

kriswest committed:
Better encapsulate handshake and identity messaging and reuse in failover handling

kriswest committed:
Adding timeouts to loaders and eliminating TimeoutLoader

Individual timeouts allow for steps that our outside the timeout (identity validation)

kriswest committed:
Limit connection strategies based on persisted data

kriswest committed:
use a previously persisted agentUrl in PostMessageLoader if available

kriswest committed:
Update packages/fdc3-get-agent/src/ui/DefaultDesktopAgentChannelSelector.ts

kriswest committed:
Dealing with review comments

kriswest committed:
Merge remote-tracking branch 'upstream/fdc3-for-web-impl' into impl-review-updates

kriswest committed:
Bridging schema fix and updating package-lock

kriswest committed:
addressing weird semgrep errors on log messages

kriswest committed:
tweakng typing in hopes of fixing a circular dependency import...

TypeError: Class extends value undefined is not a constructor or null
    at Object.<anonymous> (/home/runner/work/FDC3/FDC3/packages/fdc3-get-agent/src/ui/DefaultDesktopAgentIntentResolver.ts:11:1)

kriswest committed:
typing on waitFor and exchange + some cleanup in AbstractWebMessaging (which needs further work)

kriswest committed:
reenable code coverage

Roaders committed:
Modify generate-type-predicates to only include interfaces that have the correct type

Roaders committed:
Tidy up code. Add valid / non valid predicates

robmoffat committed:
Work in progress

robmoffat committed:
Work in progress

robmoffat committed:
Merge branch 'main' into fdc3-for-web-impl-draggable

robmoffat committed:
Drag now respects closest sides

robmoffat committed:
Fairly stable

kriswest committed:
Further cleanup of Messaging and Listeners, additional typing

ImplementationMetadata is now passed into AbstractMessaging's constructor and getting close to eliminating AbstractWebMessaging. Correcting typing and generics on a number of functions (replacing any and unqualified generics)

kriswest committed:
Correcting issues with tests (a circular dependency and unused imports)

kriswest committed:
Cleanup and simplify AbstractListener

kriswest committed:
Merge branch 'fdc3-for-web-fix-private-channel-events' into fdc3-for-web-impl-getAgent-refactor

kriswest committed:
Refactor PrivateChannel event listener support

kriswest committed:
Messaging refactor

- Eliminate AbstractWebMessaging
- Messaging is no longer a Connectable (as the connection is provided to it - was previously a no-op)
- Handling getInfo via a message exchange rather than passing down through connection flow
- Messaging implementations no longer handle implementationMetadata, instead holding the AppIdentifier directly
- Renamed HandshakeSupport HeartbeatSupport as its no longer involved in holding implementaiton metadata.
- Improving typing and error handling (defensive coding) in get-agent-proxy

kriswest committed:
Corrections to PrivateChannelAddEventListenerRequest

kriswest committed:
Merge pull request #1444 from Roaders/fdc3-for-web-impl-merging-prs

Modify generate-type-predicates to only include interfaces that have …

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl-update-type-predicates

kriswest committed:
Removing an unused test step with a type issue and correcting other typing issues

robmoffat committed:
Merge pull request #1447 from finos/fdc3-for-web-impl-update-type-predicates

Modify generate-type-predicates to only include interfaces that have...

robmoffat committed:
Fast failing of coverage

kriswest committed:
Merge pull request #1448 from finos/fast-fail-for-coverage

Fast failing of coverage

kriswest committed:
Merge branch 'fdc3-for-web-impl-update-type-predicates' into fdc3-for-web-impl-getAgent-refactor

kriswest committed:
More typing work on Listener and PrivateChannel events

robmoffat committed:
Restyled channel selector and got it responding to drag

robmoffat committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-draggable

robmoffat committed:
Merge pull request #1450 from finos/fdc3-for-web-impl-draggable

Fdc3 for web impl draggable

robmoffat committed:
Merge remote-tracking branch 'interopio/monorepo-docusaurus' into fdc3-for-web-impl-docusaurus

robmoffat committed:
Fixing netlify build

robmoffat committed:
Fixing netlify

robmoffat committed:
Removed yarn.lock to enable netlify to use npm

robmoffat committed:
fixing mssing dev dep

robmoffat committed:
adding install step in website build

robmoffat committed:
fixing netlify build

robmoffat committed:
Merge pull request #1451 from finos/fdc3-for-web-impl-docusaurus

Fdc3 for web impl docusaurus

kriswest committed:
Better logging in getAgent

robmoffat committed:
Added legacy CJS support fdc3 package

robmoffat committed:
removed command line args, added to script

robmoffat committed:
removed comments

robmoffat committed:
Added display block/none on Intent resolver iframe

kriswest committed:
Send WCP6Goodbye on disconnects - inside MessagePortMessaging

kriswest committed:
minor tweaks to heartbeats

kriswest committed:
typing in UI component handling

kriswest committed:
Add fdc3 version to standard package

kriswest committed:
Additional logging in getAgent

kriswest committed:
minor logging tweaks

kriswest committed:
minor schema correction in fdc3UserInterfaceResolveAction

kriswest committed:
iframe adaptor responds to hello message

kriswest committed:
Refactoring fdc3-get-agent tests

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-getAgent-refactor

kriswest committed:
getAgent tests passing + reducing log messaging during tests

kriswest committed:
some minor logging additions for debug use

robmoffat committed:
Merge remote-tracking branch 'InteropIO/impl-review-updates' into fdc3-for-web-impl-julianna-review

robmoffat committed:
Fixed interface matches (fixing tests)

robmoffat committed:
Merge pull request #1453 from finos/fdc3-for-web-iframe-display]

Fdc3 for web iframe display

robmoffat committed:
Fixed issue with channel selection callback|

robmoffat committed:
Merge pull request #1454 from finos/fdc3-for-web-impl-julianna-review

Fdc3 for web impl julianna review

kriswest committed:
adding prettier back in

kriswest committed:
Applying prettier to all code, setting up commit hook and eslint configs

kriswest committed:
adding linting scripts

kriswest committed:
improving typescript use in fdc3-agent-proxy

kriswest committed:
adjusting privateChannel tests to match correct to private channel event types

kriswest committed:
code formating and minor improvements in toolbox/fdc3-for-web implementation

kriswest committed:
Improved typing and fixes to agent-proxy tests (6 still need fixing)

kriswest committed:
agent-proxy tests passing, fix PrivateChannel event and getInfo handling

kriswest committed:
dealing private channel event listener changes in fdc3-web-impl

kriswest committed:
Adding linting to toolbox/fdc3-web-impl projects

kriswest committed:
more typing and formatting in fdc3-get-agent

kriswest committed:
more test cleanup and adding a test for failed validation

kriswest committed:
fdc3-get-agent tests coverage

robmoffat committed:
Testing husky lint

robmoffat committed:
Setting fetch depth

robmoffat committed:
adding fetch base

kriswest committed:
Further refactor of fdc3-get-agent tests and improving coverage

robmoffat committed:
Merge pull request #1456 from InteropIO/fdc3-web-impl-prettier

Fdc3 web impl prettier + eslint

kriswest committed:
Merging eslint and prettier changes

kriswest committed:
Attempting to deal with SemGrep issues with log messages

kriswest committed:
Fixing remaining BrowserTypes imports

kriswest committed:
Further work on resolving logging semgrep complaints

kriswest committed:
Further attempt to resolve SemGrep logging issues

kriswest committed:
reworking Logger again to appease semgrep

kriswest committed:
Output error stack traces in Logger

kriswest committed:
failoverhandler test coverage

kriswest committed:
lint free fdc3-get-agent

kriswest committed:
lint in fdc3-web-impl

kriswest committed:
lint in fdc3-web-impl and reference-ui

kriswest committed:
lint in demo

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl

kriswest committed:
Handling private channel schema change

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-getAgent-refactor

robmoffat committed:
Update coverage.yml

kriswest committed:
Addressing review feedback

kriswest committed:
Removing redundant cleanup steps

kriswest committed:
Clear timers when no longer needed in AbstractMessaging

robmoffat committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-commonjs

robmoffat committed:
build complete

kriswest committed:
Update function in docs that searches for parent frames

kriswest committed:
Renaming BasicDesktopAgent to DesktopAgentProxy (as that better represents what it does)

kriswest committed:
minor formatting spelling and typing fixes in fdc3-agent-proxy

robmoffat committed:
Added commonjs as a separate module

robmoffat committed:
Updated package lock

kriswest committed:
Address review comments

robmoffat committed:
Update README.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Added module readmes

robmoffat committed:
added changelog

kriswest committed:
Resolving comments from Rob's review

robmoffat committed:
Update coverage.yml

robmoffat committed:
reverted

kriswest committed:
Remove defunct code in DefaultIntentListener

kriswest committed:
Fix error introduced into UI iframe test infrastructure

kriswest committed:
Resolving review comments from Julianna's final batch

robmoffat committed:
Refactored FailoverHandler into separate smaller functions, tidied up logging tests so it's clearer what they're doing

robmoffat committed:
Merge branch 'fdc3-for-web-impl-getAgent-refactor' into fdc3-for-web-impl-getAgent-refactor-rm-patch1

robmoffat committed:
Adding base coverage

robmoffat committed:
Merge pull request #1477 from finos/fdc3-for-web-impl-base-coverage

Fdc3 for web impl base coverage

robmoffat committed:
Update base-coverage.yml

kriswest committed:
Merge pull request #1476 from finos/fdc3-for-web-impl-getAgent-refactor-rm-patch1

Refactored FailoverHandler into separate smaller functions, tidied up…

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-getAgent-refactor

kriswest committed:
Note on wtfnode use

robmoffat committed:
Update packages/fdc3/README.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Update packages/fdc3/README.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Update packages/fdc3-commonjs/README.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Update packages/fdc3-commonjs/README.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Update base-coverage.yml

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl

robmoffat committed:
Merge pull request #1452 from finos/fdc3-for-web-impl-commonjs

Fdc3 for web impl commonjs

robmoffat committed:
trying to fix coverage reporting.

robmoffat committed:
changing path

kriswest committed:
Update packages/fdc3-agent-proxy/src/intents/DefaultIntentSupport.ts

robmoffat committed:
Added test policy in contributing section

robmoffat committed:
added changelog

kriswest committed:
Resolving comments from Rob's review

kriswest committed:
Addressing Rob's coverage comments

kriswest committed:
SSimplify DefaultIntentSupport.raiseIntent* coverage

kriswest committed:
Test channel selector callback

kriswest committed:
Make sure MessagePort is started for identity validation

kriswest committed:
fix dependency mismatches

Roaders committed:
migrate workbench to getAgent

kriswest committed:
Correcting issues with reconnection in fdc3-for-web-impl and demo

Reconnections (e.g. on refresh) were being rejected as the instance UUIDs Were delted on disconnect. Adjusted the agent to cache past connections for upto N seconds (currently set to 60). Once reconnection was allowed, double broadcasts and other state was observed as these were not being cleaned up on disconnect, which was also corrected.

kriswest committed:
IntentHandler test coverage

kriswest committed:
catch cross-domain window.name accesses

kriswest committed:
some lint and logging

Roaders committed:
updated changelog

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-getAgent-refactor

kriswest committed:
cleanup package-lock

kriswest committed:
Better validation of addContextListener args and testing of error on invalid args

kriswest committed:
Revert to importing all of BrowserTypes as CommonJS compile doesn't like them

kriswest committed:
Addressing review comments including using appState.terminated for past connections

kriswest committed:
Refining fdc3-for-web-impl's findIntent fns, tests and cleanup on disconnection

kriswest committed:
rolling package verison numbers

kriswest committed:
Fix getAgent tests after previous change to fdc3-for-web-impl

kriswest committed:
syncpack

kriswest committed:
syncing more package.json version nums

kriswest committed:
Import picocolors to fix commonjs build

n.b. Chalk also runs into issues due to our dual-build - may replace with a custom implementation later.

kriswest committed:
timing improvements to injected UIs

kriswest committed:
Ignore untested parts of picocolors for coverage

kriswest committed:
correcting updateChannel in test implementations of channelSelectors

kriswest committed:
Picocolors testing exemption

kriswest committed:
protect messaging in DefaultChannel

kriswest committed:
Adding notes on non-standard error messages

kriswest committed:
Revert dummy-desktop-agent to default directory

kriswest committed:
Improving channel selector implementations

kriswest committed:
Remove picocolors

kriswest committed:
remove unused glyphColor

robmoffat committed:
Merge pull request #1466 from finos/fdc3-for-web-impl-getAgent-refactor

Fdc3 for web impl get agent refactor

robmoffat committed:
Update CONTRIBUTING.md

robmoffat committed:
Update CONTRIBUTING.md

robmoffat committed:
Update CONTRIBUTING.md

Co-authored-by: Kris West <kris.west@interop.io>

robmoffat committed:
Moved from kite9 -> finos

robmoffat committed:
2.2.0-alpha.1

robmoffat committed:
Updated version number, published 2.2.0-alpha.1

robmoffat committed:
removed spurious comments

robmoffat committed:
Added documentation

robmoffat committed:
Fixing image paths

kriswest committed:
Adding support for controlling debug logging in fdc3-agent-proxy and fdc3-get-agent

kriswest committed:
CHANGELOG

kriswest committed:
Testing coverage

kriswest committed:
Test coverage

kriswest committed:
Merge pull request #1480 from finos/fdc3-for-web-impl-test-policy

Fdc3 for web impl test policy

robmoffat commented:
@kriswest I think you're going to need to go through EasyCLA ⬆️

robmoffat committed:
Added dist, bumped to alpha.2 version

kriswest commented:
@robmoffat doh, I wonder why those commits are not tied to my github id :-( I'll work on this later on today 😭

kriswest committed:
Allow DAs to customize timeouts used for message exchanges and app launches

kriswest committed:
timeouts in testing for fdc3-agent-proxy

kriswest committed:
Shorten timeouts used when testing

kriswest commented:
/easycla

robmoffat committed:
updated to alpha.3

kriswest committed:
Update website/docs/api/ref/GetAgent.md

kriswest committed:
WIP timeout tests - not passing yet

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl

kriswest committed:
Apply prettier after generating files from schemas

At various times we've committed the generated source files with or without prettier reformatting them. Always apply prettier after generation to make them more stable.

Roaders commented:
Tested against our Desktop Agent using Workbench from PR #1485 (rebased onto branch `fdc3-for-web-impl`)

All workbench functionality successfully tested:

- addContextListener
- addIntentListener
- broadcast
- getCurrentChannel
- getInfo
- getOrCreateChannel
- getUserChannels
- joinUserChannel
- leaveCurrentChannel
- raiseIntent
- raiseIntentForContext

raiseIntent / raiseIntentForContext tested without UI Iframes enabled in desktop agent. App Resolution was handled in container window.

Untested:

- createPrivateChannel
- findInstances
- findIntent
- findIntentsByContext
- getAppMetadata
- open

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-debug-logs

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl

kriswest committed:
Migrate conformance test docs into correct location in mono repo

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-debug-logs

kriswest committed:
Merge branch 'main' into fdc3-for-web-impl-control-timeouts

kriswest committed:
Move conformance test docs into correct location in mono-repo

robmoffat committed:
Fixing packaging and deployment of Reference UI for Web

robmoffat committed:
Using hosted channel selector and intent resolver.

robmoffat commented:
ok, I've tested UI hosted on the preview site.  There's now a third option in the demo dropdown to allow you to choose which default UI you want to use:

![Screenshot 2025-01-17 at 12 35 55](https://github.com/user-attachments/assets/5d8d4ff2-5b63-4ab2-bc18-dcf6139cf86b)

When we merge this PR, we need to change this one spot in `util.ts`:

```
  [UI.DEFAULT]: {
    /* REPLACE THIS URL WITH fdc3.finos.org BEFORE RELEASE */
    intentResolverUrl: 'https://deploy-preview-1309--fdc3.netlify.app/toolbox/fdc3-reference-ui/intent_resolver.html',
    channelSelectorUrl: 'https://deploy-preview-1309--fdc3.netlify.app/toolbox/fdc3-reference-ui/channel_selector.html',
  },
```

robmoffat committed:
Remove unused files originally written for early development

robmoffat committed:
fixed dup in test policy

kriswest committed:
Creating ApiTimeout errors and partially complete tests

kriswest committed:
Merge branch 'fdc3-for-web-impl-debug-logs' into fdc3-for-web-impl-control-timeouts

kriswest committed:
first test working

kriswest committed:
Log handshake messages at debug level

kriswest committed:
All timeout tests working

kriswest committed:
CHANGELOG

kriswest committed:
Merge branch 'fdc3-for-web-impl' into fdc3-for-web-impl-control-timeouts

kriswest committed:
restoring missing test step

kriswest committed:
Minor schema correction

robmoffat commented:
/netlify

kriswest committed:
switch existing test to new error message (as its an APITimeout rather than the app timing it - Desktop Agent should report IntentDelvieryFailed itself)

Roaders committed:
Update CHANGELOG.md

Co-authored-by: Kris West <kris.west@interop.io>

kriswest committed:
move non-generic test steps to their own file in fdc3-get-agent and delete a redundant step

kriswest committed:
defaultTimeout -> messageExchangeTimeout

kriswest committed:
defaultTimeout -> messageExchangeTimeout (in tests)

robmoffat commented:
Preview here: https://deploy-preview-1309--fdc3.netlify.app

kriswest committed:
Rework logging config based on log levels

robmoffat committed:
Fixed path to directory schema

robmoffat committed:
Added extra exports

kriswest committed:
Log settings in tests + coverage

robmoffat committed:
Merge pull request #1485 from Roaders/fdc3-for-web-impl-getAgent-refactor

Migrate workbench to getAgent

robmoffat committed:
fixed typo

kriswest committed:
Correcting NONE log level definition

kriswest committed:
Cleanup of an unused type and incorrect timeout in a test

kriswest committed:
Clarifying note on timeouts in DACP docs

robmoffat committed:
Merge pull request #1500 from finos/fdc3-for-web-impl-fixes-rm

Fdc3 for web impl fixes rm

robmoffat committed:
Merge pull request #1497 from finos/fdc3-for-web-impl-control-timeouts

Allow DAs to customize timeoutsand apps to control logging

kriswest reviewed (dismissed):
I've spent an enormous amount of time with this PR and the changes it makes to the repository, new code implemented, documentation improvments etc. and am satisfied that it LGTM!

kriswest committed:
fix linting config in fdc3-workbench

kriswest reviewed (dismissed):
LGTM again  (after fixing fdc3-workbench linter config)

novavi reviewed (dismissed):
Please see comments for issues I found.

robmoffat commented on toolbox/fdc3-for-web/demo/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@finos/demo",
+  "private": true,
```
yes again

robmoffat commented on toolbox/fdc3-for-web/reference-ui/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "fdc3-for-web-reference-ui",
+  "private": true,
```
yes - this isn't published into npm, it goes in the website build

robmoffat commented on toolbox/fdc3-for-web/reference-ui/package.json:
```diff
@@ -0,0 +1,27 @@
+{
```
do we need these if it's not published?

bingenito replied:
My personal opinion is no

kriswest commented:
@robmoffat are you ok to own resolving @novavi's comments? I shouldn't push currently as that will invalidate review and cause some CLA issues)

kriswest commented on toolbox/fdc3-for-web/reference-ui/eslint.config.js:
```diff
@@ -0,0 +1,13 @@
+import globals from 'globals';
```
Yes probably

bingenito replied:
@robmoffat Can you please look at this one as I don't know how to easily test/confirm the change impact.

robmoffat replied:
I changed it and ran it, it still works.

kriswest commented on toolbox/fdc3-workbench/package.json:
```diff
@@ -1,9 +1,8 @@
   "name": "fdc3-workbench",
-  "version": "2.0.0",
+  "version": "2.2.0-alpha.3",
   "private": true,
```
I believe so - all it does is ensure that NPM will refuse to publish the module. This was intended to be hosted or built from source rather than pulled from NPM

kriswest commented on toolbox/fdc3-for-web/reference-ui/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "fdc3-for-web-reference-ui",
+  "private": true,
```
Good point

bingenito replied:
Updated

kriswest commented on toolbox/fdc3-for-web/demo/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@finos/demo",
+  "private": true,
```
Good point

bingenito replied:
Updated

kriswest commented on website/docs/api/specs/webConnectionProtocol.md:
Yes it should

bingenito replied:
Updated

kriswest commented on website/docs/api/specs/desktopAgentCommunicationProtocol.md:
Yes it should be ../../

bingenito replied:
Updated

kriswest commented on website/docs/api/specs/browserResidentDesktopAgents.md:
Yes it should be ../../

kriswest commented on website/docs/api/specs/preloadDesktopAgents.md:
Agreed

bingenito replied:
Updated

kriswest commented on website/docs/api/specs/webConnectionProtocol.md:
Agreed

bingenito replied:
Updated

bingenito commented:
> @robmoffat are you ok to own resolving @novavi's comments? I shouldn't push currently as that will invalidate review and cause some CLA issues)

I can go through a bunch and add these tomorrow.  It can be a good way to help participate in this PR as I'm limited in what I can review.

robmoffat commented:
> > @robmoffat are you ok to own resolving @novavi's comments? I shouldn't push currently as that will invalidate review and cause some CLA issues)
> 
> I can go through a bunch and add these tomorrow. It can be a good way to help participate in this PR as I'm limited in what I can review.

awesome thanks Brian

bingenito committed:
Fix capitalization of npm

bingenito committed:
Add and update path references to DesktopAgent in docs

bingenito committed:
Fix relative doc path references to fdc3-compliance

bingenito committed:
Add missing license identifier

bingenito committed:
Correct standard package metadata such as homepage

bingenito commented on toolbox/fdc3-workbench/package.json:
```diff
@@ -1,9 +1,8 @@
 {
```
Updated

bingenito commented on toolbox/fdc3-for-web/reference-ui/package.json:
```diff
@@ -0,0 +1,27 @@
+  "name": "fdc3-for-web-reference-ui",
+  "private": true,
+  "version": "2.2.0-alpha.3",
+  "type": "module",
```
@robmoffat Can you please look at this one as I don't know how to easily test.

robmoffat committed:
Renamed lint config file.

kriswest reviewed (approved):
LGTM

robmoffat commented:
@novavi can you please check @bingenito's commits fix your outstanding review issues and we can get this merged?

thanks,
Rob
